### PR TITLE
Retain Payment Information in API DB (Payment History)

### PR DESCRIPTION
**Pull request by shadowcpy**
Implementing #569
This PR will enable the CLI and third-party clients to display a TX-History by retaining information about every transaction

### CODE REVIEW

**Comment by justinmoon** [comment]
One thing that we should probably do here (or perhaps elsewhere) is encode fees taken by the gateway more explicitly

File: client/client-lib/src/ln/db.rs [hunk 1]
```diff
@@ -93,3 +97,22 @@ impl DatabaseKeyPrefixConst for LightningGatewayKey {
     type Key = Self;
     type Value = LightningGateway;
 }
+
+#[derive(Debug, Clone, Encodable, Decodable)]
+pub struct PaymentKey(pub sha256::Hash);
```
Comment (elsirion) [inline client/client-lib/src/ln/db.rs:102]: This key should contain some sort of timestamp as the first field. Our databse is essentially an on-disk `BTreeMap`, so having the thing we (presumably?) want to sort by firs tin the key is very useful.

File: client/client-lib/src/lib.rs [hunk 2]
```diff
@@ -111,6 +112,70 @@ impl From<GatewayClientConfig> for LightningGateway {
...
+        contract_id: ContractId,
+        used_gateway: PublicKey,
+    },
+    InboundMint {
+        used_token: TieredMulti<SpendableNote>,
+        reissued_token: Option<TieredMulti<SpendableNote>>,
+    },
+    OutboundMint {
+        token: TieredMulti<SpendableNote>,
+    },
```
Comment (elsirion) [inline client/client-lib/src/lib.rs:148]: These will probably never be used but still cool to have them

File: client/client-lib/src/lib.rs [hunk 3]
```diff
@@ -619,14 +685,36 @@ impl Client<UserClientConfig> {
...
+                    saved_payment.detail = TypedPaymentDetail::InboundLightning {
+                        invoice: (invoice),
+                        paid: true,
+                        used_gateway: (used_gateway),
+                    }
+                }
+                _ => panic!("retrieved payment history is of wrong type! (LN)"),
+            }
+            self.ln_client().save_payment(&saved_payment);
+        }
```
Comment (elsirion) [inline client/client-lib/src/lib.rs:705]: Updating of redundant data seems risky. We might forget to update payments when e.g. introducing a new failure case. I'd much rather only save the payment details once and then use other DB entries to determine payment state (even though that's a bit less efficient, but we'd paginate anyway I assume).

File: client/client-lib/src/lib.rs [hunk 4]
```diff
@@ -111,6 +112,70 @@ impl From<GatewayClientConfig> for LightningGateway {
...
+        note: Option<String>,
+        detail: TypedPaymentDetail,
+    ) -> Self {
+        Payment {
+            uuid: (Uuid::new_v4()),
+            timestamp: (duration_since_epoch().as_millis()),
+            amount: (amount),
+            direction: (direction),
+            note: (note),
+            detail: (detail),
```
Comment (elsirion) [inline client/client-lib/src/lib.rs:174]: Why not use the shorter notation here?

File: client/client-lib/src/lib.rs [hunk 5]
```diff
@@ -111,6 +112,70 @@ impl From<GatewayClientConfig> for LightningGateway {
...
+        token: TieredMulti<SpendableNote>,
+    },
+}
+
+#[derive(Clone, Debug, Encodable, Decodable, Serialize, Deserialize)]
+pub struct Payment {
+    pub uuid: Uuid,
+    pub timestamp: u128,
+    pub amount: Amount,
+    pub direction: PaymentDirection,
```
Comment (elsirion) [inline client/client-lib/src/lib.rs:156]: The direction is already included in the `detail` field imo, couldn't `TypedPaymentDetail` have a function `direction(&self) -> PaymentDirection` instead?

File: client/client-lib/src/ln/db.rs [hunk 6]
```diff
@@ -12,6 +15,7 @@ const DB_PREFIX_OUTGOING_PAYMENT_CLAIM: u8 = 0x24;
 const DB_PREFIX_OUTGOING_CONTRACT_ACCOUNT: u8 = 0x25;
 const DB_PREFIX_CONFIRMED_INVOICE: u8 = 0x26;
 const DB_PREFIX_LIGHTNING_GATEWAY: u8 = 0x28;
+const DB_PREFIX_PAYMENTS: u8 = 0x29;
```
Comment (elsirion) [inline client/client-lib/src/ln/db.rs:18]: requires an update of the [DB docs](https://github.com/fedimint/fedimint/blob/master/docs/database.md)

**Comment by elsirion** [comment]
Thx for tackling this!

**Pull request closed without merging**